drivers/rtc/rtc_ll_stm32.c

```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+			p_rtc_alrm->user_mask = RTC_ALARM_TIME_MASK_SECOND;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_MINUTES) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MINUTE;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_HOURS) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_HOUR;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_DATEWEEKDAY) == 0x0) {
+			if (timeptr->tm_wday) {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_WEEKDAY;
+			} else {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+			}
+		}
+	}
+	*mask = p_rtc_alrm->user_mask;
+
+	LOG_DBG("get alarm: mday = %d, wday = %d, hour = %d, min = %d, sec = %d, "
+		"mask = 0x%04x", timeptr->tm_mday, timeptr->tm_wday, timeptr->tm_hour,
+		timeptr->tm_min, timeptr->tm_sec, *mask);
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int rtc_stm32_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+			const struct rtc_time *timeptr)
+{
+	struct rtc_stm32_data *data = dev->data;
+	struct rtc_stm32_alrm *p_rtc_alrm;
+	LL_RTC_AlarmTypeDef *p_ll_rtc_alarm;
+	LL_RTC_TimeTypeDef *p_ll_rtc_alrm_time;
+	uint32_t ll_mask = 0;
+	int err = 0;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	if (id == RTC_STM32_ALRM_A) {
+		p_rtc_alrm = &(data->rtc_alrm_a);
+	} else if (id == RTC_STM32_ALRM_B) {
+		p_rtc_alrm = &(data->rtc_alrm_b);
+	} else {
+		LOG_ERR("invalid alarm ID %d", id);
+		err = -EINVAL;
+		goto unlock;
+	}
+
+	if ((mask == 0) && (timeptr == NULL)) {
+		memset(&(p_rtc_alrm->ll_rtc_alrm), 0, sizeof(LL_RTC_AlarmTypeDef));
+		p_rtc_alrm->user_callback = NULL;
+		p_rtc_alrm->user_data = NULL;
+		p_rtc_alrm->is_pending = false;
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+		LL_PWR_EnableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPCR_DBP || PWR_DBPR_DBP */
+		if (ll_func_is_active_alarm(RTC, id)) {
+			LL_RTC_DisableWriteProtection(RTC);
+			ll_func_disable_alarm(RTC, id);
+			LL_RTC_EnableWriteProtection(RTC);
+		}
+		LOG_DBG("Alarm %d has been disabled", id);
+		goto disable_bkup_access;
+	}
+
+	if ((mask & ~(p_rtc_alrm->supported_mask)) != 0) {
+		LOG_ERR("unsupported alarm %d field mask 0x%04x", id, mask);
+		err = -EINVAL;
+		goto unlock;
+	}
+
+	if (timeptr == NULL) {
+		LOG_ERR("timeptr is invalid");
+		err = -EINVAL;
+		goto unlock;
+	}
+
+	if (p_rtc_alrm->user_callback != NULL) {
+		LOG_ERR("Alarm %d is busy", id);
+		err = -EBUSY;
+		goto unlock;
+	}
+
+	p_ll_rtc_alarm = &(p_rtc_alrm->ll_rtc_alrm);
+	p_ll_rtc_alrm_time = &(p_ll_rtc_alarm->AlarmTime);
+
+	memset(p_ll_rtc_alrm_time, 0, sizeof(LL_RTC_TimeTypeDef));
+
+	/*
+	 * RTC Alarm LL mask should be set for all fields beyond the biggest one
+	 * that's being matched with RTC calendar to trigger alarm periodically
+	 */
+	if (id == RTC_STM32_ALRM_A) {
+		ll_mask = LL_RTC_ALMA_MASK_ALL;
+	} else {
+		ll_mask = LL_RTC_ALMB_MASK_ALL;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		if (timeptr->tm_sec > 59) {
```

Comment:
Have you seen 
https://github.com/zephyrproject-rtos/zephyr/blob/35429debb627943410f3c3651f267fb40f04a9cf/drivers/rtc/rtc_utils.c#L15C6-L15C33 ? unless the alarms have some additional constraint on the time fields, it can save a bit of repeated code :)